src/Take.Blip.Builder/FlowManager.cs

```diff
@@ -266,14 +263,15 @@ public async Task ProcessInputAsync(Message message, Flow flow, CancellationToke
             }
             catch (Exception ex)
             {
-                _logger.Error(ex, "Error processing input '{Input}' for user '{User}'", message.Content, userIdentity);
-
                 if (inputTrace != null)
                 {
                     inputTrace.Error = ex.ToString();
                 }
-                
-                throw;
+
+                throw new BuilderStateException($"Error processing input '{message.Content}' for user '{userIdentity}' in state '{state?.Id}'", ex)
+                {
+                    StateId = state.Id
+                };
```

Comment:
I think that we only need 3 exception types here: 
* BuilderException, with StateId, UserIdentity and maybe MessageId properties
* OutputProcessingException, inheriting from BuilderException and adding OutputStateId and OutputConditions as you defined (you can fill BuilderException properties from LazyInput and IContext)
* ActionProcessingException, inheriting from BuilderException and adding ActionType and ActionSettings as you defined (you can fill BuilderException properties from LazyInput and IContext)

In this exception handling code, you could just catch first for a BuilderException (which will get the previous too) and then make the inputtrace test and just throw. Then another catch for Exception, where you will wrap all other exceptions inside a BuilderExcetion, with its properties filled.